### PR TITLE
wip: add applet module

### PR DESCRIPTION
**Pull request by wash2**
This seems to address the issues from https://github.com/pop-os/cosmic-applets/pull/110 but maybe a separate trait could be added which adds the style by default.

### CODE REVIEW

File: src/app/applet/mod.rs [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+            parent_size: None,
+            grab: true,
+        }
+    }
+
+    #[must_use]
+    pub fn theme(&self) -> theme::Theme {
+        match self.background {
+            CosmicPanelBackground::ThemeDefault | CosmicPanelBackground::Color(_) => {
+                let Ok(helper) = cosmic_config::Config::new(
```
Comment (ids1024) [inline src/app/applet/mod.rs:205]: Don't we already have a `system_theme` that should contain this?

Comment (wash2) [reply src/app/applet/mod.rs:205]: ya i'll update this

File: cosmic-theme/Cargo.toml [hunk 2]
```diff
@@ -12,21 +12,16 @@ rustdoc-args = ["--cfg", "docsrs"]
...
-hex-color = ["hex"]
+theme-from-image = ["kmeans_colors", "image"]
 
 [dependencies]
-palette = {version = "0.7", features = ["serializing"] }
-anyhow = "1.0"
-hex = {version = "0.4.3", optional = true}
+# palette = {version = "0.7", features = ["serializing"] }
+almost = "0.2"
+palette = {git = "https://github.com/Ogeon/palette", features = ["serializing"] }
```
Comment (ids1024) [inline cosmic-theme/Cargo.toml:20]: Palette 0.7.3 was just released. I presume that has what we need?

File: Cargo.toml [hunk 3]
```diff
@@ -97,3 +107,7 @@ exclude = [
 
 [patch."https://github.com/pop-os/libcosmic"]
 libcosmic = { path = "./", features = ["wayland", "tokio", "a11y"]}
+
+# TODO Remove me when the palette crate gets an update & before merging
+[patch.crates-io]
+palette = {git = "https://github.com/Ogeon/palette", features = ["serializing"] }
```
Comment (ids1024) [inline Cargo.toml:113]: A `patch` section only works in a top-level crate, so if this is used as a dependency by something else (like cosmic-applets) the patch doesn't apply and it fails to build.

Comment (wash2) [reply Cargo.toml:113]: Oh, ya, that needs to be removed. It was removed from the theme update branch but not this one.

**Comment by ids1024** [comment]
This actually wasn't compiling with the `applet` feature, so I added a commit that fixed two little things that were breaking it.

Using this in an applet, it seems like https://github.com/pop-os/libcosmic/commit/20a5227eca5d566427ecdfdb132ec9bd4becc2f1 has changed the styling for the applet buttons and buttons within them so the `hover` state isn't (very) visibly different. Changing `Button::Text => &cosmic.text_button` back to `Button::Text => &theme.current_container().component` reverts to the previous styling. But I guess that's something we want to change in the default theme instead?

**Comment by ids1024** [comment]
Otherwise this is looking good. I wonder if `style()` could be handled automatically somehow when `applet::run` is used to avoid that (small) boilerplate. And then there's the need to create a `CosmicAppletHelper` while the code in `cosmic::app::applet` is already doing that, though that was already duplicated.

**Comment by wash2** [comment]
> And then there's the need to create a `CosmicAppletHelper` while the code in `cosmic::app::applet` is already doing that, though that was already duplicated.

I think applets can just use `self.core.applet_helper` now instead of creating an `CosmicAppletHelper`, or am I missing something?

**Comment by ids1024** [comment]
Ah right, it's a property of `core`. So yeah, that part is simple enough.

**Comment by ids1024** [comment]
It doesn't seem quite right with the last commit either. The power applet here is styled using this branch, while the other applets are old versions.

eDP-1

Presumably when the applet isn't focused we still want the background to match the panel or be transparent.

**Comment by wash2** [comment]
> It doesn't seem quite right with the last commit either. The power applet here is styled using this branch, while the other applets are old versions.

Ahh ok, I see what you mean. I'll just make it use a custom style then?

**Comment by mmstick** [comment]
Perhaps the hover state visibility could be a property of the Appearance struct, or as an input parameter to our methods?

**Comment by ids1024** [comment]
Looks like the appearance when not hovered is still like my last screenshot. With a somewhat darker circle for the button instead of matching the panel.

Previously this used a completely transparent background for the button when not hovered, and an opaque background when hovered. So I guess we need that to get the same behavior.

A over A = A for opaque colors, but for semi-transparent colors it gets the same color but with a higher alpha.

**Comment by wash2** [comment]
> Looks like the appearance when not hovered is still like my last screenshot. With a somewhat darker circle for the button instead of matching the panel.
> 
> Previously this used a completely transparent background for the button when not hovered, and an opaque background when hovered. So I guess we need that to get the same behavior.
> 
> A over A = A for opaque colors, but for semi-transparent colors it gets the same color but with a higher alpha.

Oh sorry I forgot to make the buttons text buttons again after making the fix. Backgrounds of text buttons should be completely transparent.

**Review by ids1024 — APPROVED** [review]
I added a commit to not have "applet" depend on "a11y" since that was causing a panic... hopefully we can get that working well later.

Seems to be working fairly well now.

**Pull request merged**